tests/System.Buffers.Primitives.Tests/CtorArray.cs

```diff
@@ -0,0 +1,216 @@
+            Buffer<long> buffer;
+
+            buffer = new Buffer<long>(a);
+            TestHelpers.Validate(buffer, 91, -92, 93, 94, -95);
+
+            buffer = new Buffer<long>(a, 0);
+            TestHelpers.Validate(buffer, 91, -92, 93, 94, -95);
+
+            buffer = new Buffer<long>(a, 0, a.Length);
+            TestHelpers.Validate(buffer, 91, -92, 93, 94, -95);
+        }
+
+        [Fact]
+        public static void CtorArray3()
+        {
+            object o1 = new object();
+            object o2 = new object();
+            object[] a = { o1, o2 };
+            Buffer<object> buffer;
+
+            buffer = new Buffer<object>(a);
+            TestHelpers.Validate(buffer, o1, o2);
+
+            buffer = new Buffer<object>(a, 0);
+            TestHelpers.Validate(buffer, o1, o2);
+
+            buffer = new Buffer<object>(a, 0, a.Length);
+            TestHelpers.Validate(buffer, o1, o2);
+        }
+
+        [Fact]
+        public static void CtorArrayZeroLength()
+        {
+            int[] empty = Array.Empty<int>();
+            Buffer<int> buffer;
+
+            buffer = new Buffer<int>(empty);
+            TestHelpers.Validate(buffer);
+
+            buffer = new Buffer<int>(empty, 0);
+            TestHelpers.Validate(buffer);
+
+            buffer = new Buffer<int>(empty, 0, empty.Length);
+            TestHelpers.Validate(buffer);
+        }
+
+        [Fact]
+        public static void CtorArrayNullArray()
+        {
+            Assert.Throws<ArgumentNullException>(() => new Buffer<int>(null));
+            Assert.Throws<ArgumentNullException>(() => new Buffer<int>(null, 0));
+            Assert.Throws<ArgumentNullException>(() => new Buffer<int>(null, 0, 0));
+        }
+
+        [Fact]
+        public static void CtorArrayWrongArrayType()
+        {
+            // Cannot pass variant array, if array type is not a valuetype.
+            string[] a = { "Hello" };
+            Assert.Throws<ArrayTypeMismatchException>(() => new Buffer<object>(a));
+            Assert.Throws<ArrayTypeMismatchException>(() => new Buffer<object>(a, 0));
+            Assert.Throws<ArrayTypeMismatchException>(() => new Buffer<object>(a, 0, a.Length));
+        }
+
+        [Fact]
+        public static void CtorArrayWrongValueType()
+        {
+            // Can pass variant array, if array type is a valuetype.
+
+            uint[] a = { 42u, 0xffffffffu };
+            int[] aAsIntArray = (int[])(object)a;
+            Buffer<int> buffer;
+
+            buffer = new Buffer<int>(aAsIntArray);
+            TestHelpers.Validate(buffer, 42, -1);
+
+            buffer = new Buffer<int>(aAsIntArray, 0);
+            TestHelpers.Validate(buffer, 42, -1);
+
+            buffer = new Buffer<int>(aAsIntArray, 0, aAsIntArray.Length);
+            TestHelpers.Validate(buffer, 42, -1);
+        }
+
+        [Fact]
+        public static void CtorArrayInt1()
+        {
+            int[] a = { 90, 91, 92, 93, 94, 95, 96, 97, 98 };
```

Comment:
Many of these functions could be merged into a single test with test data for running the same test several times.

Your call though.